### PR TITLE
Bug 1786793 - part 1: Make update-l10n and update-project keep the existing file paths on firefox-android

### PR DESCRIPTION
**Pull request by JohanLorenzo**
Per the discussion in https://github.com/mozilla-l10n/android-l10n/pull/513, the implementation in https://github.com/mozilla-l10n/android-l10n-tooling/pull/47 may not be the least disruptive for the l10n team. This PR is an attempt at keeping the directory structure in https://github.com/mozilla-l10n/android-l10n identical.

### CODE REVIEW

File: src/mozxchannel/git/process.py [hunk 1]
```diff
@@ -112,7 +112,12 @@ def gather(self):
 
     def gather_repo(self, repo):
         basepath = repo.path
-        pc = TOMLParser().parse(mozpath.join(basepath, "l10n.toml"))
+        if basepath.endswith("firefox-android"):
+            # TODO: Support focus-android and fenix in the monorepo
+            l10n_toml_path = mozpath.join(basepath, "android-components", "l10n.toml")
+        else:
+            l10n_toml_path = mozpath.join(basepath, "l10n.toml")
```
Comment (JohanLorenzo) [inline src/mozxchannel/git/process.py:120]: In many places in `moxzchannel` there is a strong assumption that `l10n.toml` lives at the root of a repository. That won't be the case anymore in https://github.com/mozilla-mobile/firefox-android. There, we will have several `l10n.toml` files. One under the `android-components` folder, another one in `focus-android`, and another one in `fenix`. 

Maybe we should create a single `l10n.toml` file at the root of the monorepo too instead. That said, I don't know how different these files are today. So maybe this should be after `android-components` is moved to the monorepo.

File: src/mozxchannel/git/process.py [hunk 2]
```diff
@@ -201,7 +206,8 @@ def gather_repo(self, repo):
 
 
 def references(pc, basepath):
-    paths = ["l10n.toml"]
+    l10n_toml_relative_path = mozpath.relpath(pc.path, basepath)
+    paths = [l10n_toml_relative_path]
```
Comment (JohanLorenzo) [inline src/mozxchannel/git/process.py:211]: Speaking of assumptions about `l10n.toml`, these lines make sure we don't hardcode it here anymore.

File: src/mozxchannel/git/process.py [hunk 3]
```diff
@@ -304,10 +312,12 @@ def createWorkdir(self, contents):
...
 
     def ensureL10nToml(self, workdir):
-        includes = {
-            '{}/l10n.toml'.format(repo.target_root)
-            for repo in self.graph.repos
-        }
+        includes = set()
+        for repo in self.graph.repos:
+            # TODO Support focus-android and fenix
+            folder = "{}/android-component".format(repo.target_root) if "firefox-android" in repo.target_root else repo.target_root
```
Comment (JohanLorenzo) [inline src/mozxchannel/git/process.py]: I'll move this hardcode when adding focus-android. That will avoid the current duplication.

File: src/mozxchannel/git/process.py [hunk 4]
```diff
@@ -268,6 +274,8 @@ def handlerev(self, src_rev):
                 for p in paths:
                     if p in other_commit.tree:
                         target_path = mozpath.join(other_repo.target_root, p)
+                        # The following line is for the monorepo to sync on top of existing files
+                        target_path = target_path.replace("/firefox-android/", "/")
```
Comment (JohanLorenzo) [inline src/mozxchannel/git/process.py:280]: This is the line that transforms a path like 
`mozilla-mobile/firefox-android/android-components/.../strings.xml` into: 
`mozilla-mobile/android-components/.../strings.xml`

This way, we keep the current files paths and the Pontoon DB doesn't need to be migrated.

File: taskcluster/ci/config.yml [hunk 5]
```diff
@@ -41,6 +41,12 @@ taskgraph:
             default-repository: https://github.com/mozilla-mobile/focus-android
             default-ref: main
             type: git
+        firefoxandroid:
+            name: "Firefox Android"
+            project-regex: firefox-android
+            default-repository: https://github.com/mozilla-mobile/firefox-android
+            default-ref: ac-prep
```
Comment (JohanLorenzo) [inline taskcluster/ci/config.yml:48]: TODO: Switch to `main` once the android-components repo is officially switched over, on Monday.

File: taskcluster/ci/update-project/kind.yml [hunk 6]
```diff
@@ -36,4 +38,12 @@ jobs:
...
-
+    mozilla-mobile/firefox-android:
+        repo-prefix: firefoxandroid
+        pr-target: ac-prep  # TODO: Change to main once we migrated android-components over
+        # Note: Path to `l10n.toml` includes a subfolder because the monorepo will contain
+        # one subfolder per android product
+        #
+        # TODO: Support more than a single product
+        l10n-toml-path: 'android-l10n/mozilla-mobile/android-components/l10n.toml'
+        destination-repo: 'mozilla-mobile/firefox-android/android-components'
```
Comment (JohanLorenzo) [inline taskcluster/ci/update-project/kind.yml:49]: I don't like this hack very much, but it does the work and we manage to get string from Pontoon ported back to the `firefox-android` monorepo. For example: https://github.com/mozilla-mobile/firefox-android/pull/26

**Review by JohanLorenzo — COMMENTED** [review]
For the record, I don't like the hacks I added in this code. That said, it does the job and the tests still pass. Per the android l10n workflow described in [bug 1786793 comment 6](https://bugzilla.mozilla.org/show_bug.cgi?id=1786793#c6), android-l10n-tooling handles step 2 and step 7.

* Step 2 was tested in https://firefox-ci-tc.services.mozilla.com/tasks/bclP05JRRLm6RcBYwz12WQ/runs/1/logs/public/logs/live.log and generated https://github.com/mozilla-l10n/android-l10n/pull/516
* Step 7 was tested in [TO ADD] and generated https://github.com/mozilla-mobile/firefox-android/pull/26

I'm happy to provide any additional details. Feel free to ask! 😃

**Comment by escapewindow** [comment]
> For the record, I don't like the hacks I added in this code. That said, it does the job and the tests still pass. Per the android l10n workflow described in [bug 1786793 comment 6](https://bugzilla.mozilla.org/show_bug.cgi?id=1786793#c6), android-l10n-tooling handles step 2 and step 7.

IMO, we can decide how to proceed:

1. We land these or similar hacks, and we live with them forever (or until the next refactor). I don't know what the long term cost is, in terms of complexity / tech debt / ugliness; I may know more after I look at your patches.
2. We land these or similar hacks, and we follow up by adjusting Pontoon's DB paths and removing these hacks at a later date. That means we'll have to deal with these issues multiple times, but we can unblock the monorepo migration.
3. We block on adjusting Pontoon's DB so we don't have to land any hacks. That means we only have to deal with this once (per non-monorepo-repo), but we may need to delay the monorepo migration.

These are likely decisions that will need to be made between the mobile, releng, and l10n teams.

File: src/mozxchannel/git/process.py (hunk 1 above)
Comment (escapewindow) [reply src/mozxchannel/git/process.py:120]: I don't love the `firefox-android` hardcode, but we'll probably be changing this block with each migration, so I'm not going to block on this.

Definitely something to address as we migrate each repo -- perhaps a config file with an l10n_toml_path map, or perhaps a root-level l10n.toml (or perhaps the root-level l10n.toml tells us where to find the appropriate toml file?) or ... ?

File: src/mozxchannel/git/process.py (hunk 4 above)
Comment (escapewindow) [reply src/mozxchannel/git/process.py:280]: I don't like this hardcode either. We may, for instance, have a staging-firefox-android monorepo that we want to test on. Not blocking on this either, but let's think about what we can do to not hardcode this in the code.

File: src/mozxchannel/git/process.py (hunk 3 above)
Comment (escapewindow) [reply src/mozxchannel/git/process.py]: Should this be "android-component**s**" with an `s`?
Obligatory hardcode comment.

One thought: we could detect monorepo somehow (single-location hardcode for firefox-android or staging-firefox-android) and then create a config map, e.g.

```
path_rewrite_subdirectory = ""
if monorepo:
    path_rewrite_subdirectory = "/android-components"

folder = f"{repo.target_root}{path_rewrite_subdirectory}"
```

That could allow us to go both ways, programmatically, where the hardcodes only happen for a top-level config dict?

Also, nit: f-strings?

File: taskcluster/ci/update-project/kind.yml (hunk 6 above)
Comment (escapewindow) [reply taskcluster/ci/update-project/kind.yml:49]: Since we want 2 config items per subdir, we may want to make this a dictionary?

```
targets-or-some-other-name:
    android-components:
        l10n-toml-path: 'android-l10n/mozilla-mobile/android-components/l10n.toml'
        destination-repo: 'mozilla-mobile/firefox-android/android-components'  # should we support staging here? '{owner}/{base_repo}/android-components' ?
        l10n-path-subdirectory: '{repo}/android-components'
        some-other-config: 'how-to-remove-/firefox-android/'
    focus-android:
        ...
```

which might make all these hacks more programmatic / config-driven?

**Review by escapewindow — APPROVED** [review]
Thanks @JohanLorenzo !

Depending on the decision / deadline, we can land this PR, fix this PR to address my nits / hardcode concerns, land as-is and write follow-up fixes, or ?

Will it be confusing for future l10n/releng/mobile folk once we're fully in the monorepo and the paths in Pontoon don't match the paths in the monorepo?

File: src/mozxchannel/git/process.py (hunk 1 above)
Comment (JohanLorenzo) [reply src/mozxchannel/git/process.py:120]: Yeah, that's where I'd love have input from the l10n team. I don't know what was the original intent regarding `l10n.toml`. I see `l10n.toml` has some kind of include mechanism and a basepath defined. Maybe we could leverage that which would help habing a root-level `l10n.toml` file. That said, I don't know if this use case has ever been supported.

File: src/mozxchannel/git/process.py (hunk 4 above)
Comment (JohanLorenzo) [reply src/mozxchannel/git/process.py:280]: I agree. For sure, this line prevents testing against [1]. I originally wanted to test my changes against our staging repo first, but there were several places where this wasn't possible. For example, the android-l10n repo defines a list of repos to track[2]. So we would have to tweak this config to include our staging repo but not against `android-l10n` which only contains production-grade strings. 

Thus, we're far from being able to leverage staging-firefox-android. That's why I went with these hacks. One thing I forgot to mention earlier: these hacks are meant to be short-lived. I left more info at [3]

[1] https://github.com/mozilla-releng/staging-firefox-android
[2] https://github.com/mozilla-l10n/android-l10n/blob/24f58160fb706a8524c9f4cb27a3a1f0483d25af/config.toml
[3] https://bugzilla.mozilla.org/show_bug.cgi?id=1797507

File: taskcluster/ci/update-project/kind.yml (hunk 6 above)
Comment (JohanLorenzo) [reply taskcluster/ci/update-project/kind.yml:49]: Good idea! I'll go with a dict once we port focus-android to the monorepo.

**Review by JohanLorenzo — COMMENTED** [review]
> IMO, we can decide how to proceed:
>
> [...]
> 
> 2. We land these or similar hacks, and we follow up by adjusting Pontoon's DB paths and removing these hacks at a later date. That means we'll have to deal with these issues multiple times, but we can unblock the monorepo migration.
> 
> [...]
> 
> These are likely decisions that will need to be made between the mobile, releng, and l10n teams.

I vote for option 2. I filed [bug 1797507](https://bugzilla.mozilla.org/show_bug.cgi?id=1797507) to track this. Let me give some context to explain why this benefits the mobile, the releng, and the l10n teams. On Monday, we perform the migration. We don't want to run too many steps because things can go south and delay the migration. The hacks I implemented in this PR were tested in production (the results above were actually run on level 3 workers because I pushed this current branch on mozilla-l10n/android-l10n-tooling), so we have enough confidence for Monday. Therefore, the mobile team can start working with the new repo on Monday. _[Argument continues below]_

> Will it be confusing for future l10n/releng/mobile folk once we're fully in the monorepo and the paths in Pontoon don't match the paths in the monorepo?

In the future, we want to avoid newcomers from being surprised by these hacks. Thus, they should eventually go away, later on. @mathjazz told me a Pontoon DB migration is part of things he does regularly and need a few hours to get it performed. This means we can do it at any time the l10n team would like to. 

> Depending on the decision / deadline, we can land this PR, fix this PR to address my nits / hardcode concerns, land as-is and write follow-up fixes, or ?

I'd go with land it as-is and make changes for the next migration, when we port focus to the monorepo. I usually prefer to make things right the first time but I have a different reasoning, this time. Allow me to explain. 

I'm sorry I'm going to be direct.`mozxchannel` is the most convoluted piece of code I have ever seen. It's the first time in my career where I am unable to understand the general flow of some code by: 

 a. executing it
 b. looking at the logs - the only logs `create-l10n-branch` shows come from the command `git pull-request`, meaning I was in the dark for 99% of the lines of codes of `mozxchannel` 
 c. reading the `--help` section.
 d. and by reading the actual code

I wasn't trying to grasp each and every detail of `mozxchannel` - I just wanted to understand the general steps.

Code usually layers steps from what human do (e.g.: follow an l10n process) to what machines do (e.g.: open a file handler). So, as a developer, you can go from something tangible and go down into details as needed. The code of `src/mozxchannel/git/process.py` feels like everything is a machine-oriented detail :sweat_smile: 

So what I ended up popping a debugger at dozens of places in the code to understand what functions do and what variables are meant to represent. 

Long story short, if I wanted to make things right, I would have to refactor many things. Although, such refactors are risky and I believe the current temporary hacks are actually much less risky. That's why, I would like to land this as is, then update the hacks as needed, then remove the hacks altogether.

File: src/mozxchannel/git/process.py (hunk 3 above)
Comment (JohanLorenzo) [reply src/mozxchannel/git/process.py]: > Should this be "android-components" with an s?

Great catch! This bug actually impacted this PR: https://github.com/mozilla-l10n/android-l10n/pull/516/files#diff-ce988e742b10a22bdaf4857b59f80e15b5ebc3641948fb352d2e5386ef4d507bR10. That said, I'm not sure the impact it had in the long run. Fixed!

> One thought: we could detect monorepo somehow (single-location hardcode for firefox-android or staging-firefox-android)

Good idea! I'm not sure if I'd like to go that route. These hacks are meant to be temporary. Making bigger changes in `mozxchannel` seems risky to me. 

> Also, nit: f-strings?

The rest of the code base still uses `.format()`. I stuck to it. It won't be too much of a problem notably because this code will eventually be reverted.
